### PR TITLE
S3 Upload -- fixes #281

### LINKED ISSUE #281: Upload directly to S3
https://github.com/usgs/sciencebasepy/blob/master/demoEx2.py shows how to script uploads to S3.

@jllong-usgs -- do you mind outlining what is required? A quick glance leads me here. https://github.com/usgs/sciencebasepy/commit/5ca9f45a840900f7a27910e971253bf056ee9662 

Pretty significant change eh? 

Looking at: `upload_large_file_upload_session` why on earth did you have to implement all your own multipart upload logic? 

Am I right in thinking that this: https://search.r-project.org/CRAN/refmans/aws.s3/html/put_object.html with `multipart == TRUE` should do the trick?

### PR DESCRIPTION
**Pull request by dblodgett-usgs**

### CODE REVIEW

**Comment by dblodgett-usgs** [comment]
@wdwatkins -- would you be willing to give this a look over? 
@jllong-usgs -- I basically ripped this off sciencebasepy, if you can look through the diff and let me know if you see anything outlandish, I'd appreciate it.

File: DESCRIPTION [hunk 1]
```diff
@@ -20,7 +20,8 @@ Imports:
     jsonlite,
     curl,
     httr (>= 1.0.0),
-    methods
+    methods,
+    mime
```
Comment (dblodgett-usgs) [inline DESCRIPTION:24]: Need to check mime type -- this is a pretty prominent dependency.

File: R/authenticate_sb.R [hunk 2]
```diff
@@ -43,13 +43,39 @@ authenticate_sb = function(username, password){
...
 	
+	token_url <- pkg.env$token_url
+	
+	token <- POST(token_url, 
+								body = list(
+									client_id = pkg.env$keycloak_client_id,
+									grant_type = "password",
+									username = username,
+									password = password
+								), encode = "form")
```
Comment (dblodgett-usgs) [inline R/authenticate_sb.R:54]: Was able to reuse most of the authentication stuff to get this hacked in.

File: R/authenticate_sb.R [hunk 3]
```diff
@@ -43,13 +43,39 @@ authenticate_sb = function(username, password){
...
 	
 	invisible(h)
 }
 
+set_keycloak_env <- function(token_resp) {
+	pkg.env$keycloak_token <- jsonlite::fromJSON(rawToChar(token_resp$content))
+	
+	pkg.env$keycloak_expire <- Sys.time() + pkg.env$keycloak_token$expires_in
+}
+
```
Comment (dblodgett-usgs) [inline R/authenticate_sb.R:78]: This json parsed into an object in the active environment seemed like a reasonable way to handle this.

File: R/item_upload_files.R [hunk 4]
```diff
@@ -120,3 +120,145 @@ multi_file_body <- function(files){
...
+			message(paste("Uploading:", file))
+		
+		mimetype <- guess_type(file, unknown = 'application/octet-stream')
+		
+		cloud_upload(file, mimetype, sb_id)
+	}
+	
+	return(invisible("Success"))
+	
+}
```
Comment (dblodgett-usgs) [inline R/item_upload_files.R:160]: I'm looping for multiple files here the way we do for item_append_files, that seem ok?

File: R/item_upload_files.R [hunk 5]
```diff
@@ -120,3 +120,145 @@ multi_file_body <- function(files){
...
+
+cloud_upload <- function(file, mimetype, itemid, chunk_size_bytes = pkg.env$chunk_size_bytes,
+												 session = current_session(), status = TRUE) {
+	
+	f_size_bytes <- file.size(file)
+	f_chunks <- as.integer(f_size_bytes / chunk_size_bytes) + 1
+	
+	f_path <- paste(itemid, basename(file), sep = "/")
+	
+	gql <- httr::handle(url = pkg.env$graphql_url)
```
Comment (dblodgett-usgs) [inline R/item_upload_files.R:170]: Honestly don't know if a handle is strictly necessary here, can't hurt.

File: R/item_upload_files.R [hunk 6]
```diff
@@ -120,3 +120,145 @@ multi_file_body <- function(files){
...
+	
+	f <- file(description = file, open = "rb")
+	
+	on.exit(close(f), add = TRUE)
+	
+	part_number <- 0
+	parts_header <- list()
+	
+	if(status)
+		pb <- txtProgressBar(0, f_chunks, 1, "-", width = 50, style = 3)
```
Comment (dblodgett-usgs) [inline R/item_upload_files.R:186]: Progress bar - so simple - so clean.

File: R/item_upload_files.R [hunk 7]
```diff
@@ -120,3 +120,145 @@ multi_file_body <- function(files){
...
+	eta <- sapply(etag_payload, function(x) {
+		sprintf('{ETag: "%s", PartNumber: %i}', gsub('"', "", x$ETag), x$PartNumber)
+	})
+	
+	eta <- paste0("[", paste(eta, collapse = ","), "]")
+	
+	run_gql_query(sprintf(
+		'query { completeMultiPartUpload(object: "%s" upload_id: "%s" parts_eTags: %s) }',
+		item_str, upload_id, eta), gql)
+}
```
Comment (dblodgett-usgs) [inline R/item_upload_files.R:264]: These gql queries were finicky as hell. Probably a better way to create them.

File: R/session_renew.R [hunk 8]
```diff
@@ -55,3 +55,31 @@ session_renew = function(password, ..., username, session=current_session()){
...
+	
+	token <- httr::POST(pkg.env$token_url, body = data, encode = "form")
+	
+	if(!token$status_code == 200)
+		warning('Unable to refresh SB cloud token. Some functionality may not work.')
+	
+	set_keycloak_env(token)
+	
+	return(invisible(TRUE))
+}
```
Comment (dblodgett-usgs) [inline R/session_renew.R:85]: This works -- it's even tested (locally)

File: R/set_endpoint.R [hunk 9]
```diff
@@ -41,6 +43,10 @@ set_endpoint = function(endpoint=c("production", "development")){
 	pkg.env$url_upload_update = paste0(pkg.env$url_base, 'file/uploadAndUpdateItem/')
 	pkg.env$url_download = paste0(pkg.env$url_base, 'file/get/')
 	pkg.env$url_login = 'https://my.usgs.gov/josso/signon/usernamePasswordLogin.do'
-	
+	pkg.env$auth_server_url = paste0(pkg.env$domain, "auth")
+	pkg.env$token_url = paste0(pkg.env$domain, 
+														 "auth/realms/ScienceBase/protocol/openid-connect/token")
+  pkg.env$keycloak_client_id = "sciencebasepy"
```
Comment (dblodgett-usgs) [inline R/set_endpoint.R:49]: This is the only public client_id that was available.

File: R/item_upload_files.R (hunk 4 above)
Comment (wdwatkins) [reply R/item_upload_files.R:161]: Yeah, wouldn't be much use in paralellizing

File: R/session_renew.R [hunk 10]
```diff
@@ -55,3 +55,31 @@ session_renew = function(password, ..., username, session=current_session()){
 		invisible(authenticate_sb(sb_username, password))
 	}
 }
+
+
+refresh_token_before_expired <- function(refresh_amount_seconds = 600) {
```
Comment (wdwatkins) [inline R/session_renew.R:60]: How long do the tokens generally last?

File: R/item_upload_files.R (hunk 7 above)
Comment (wdwatkins) [reply R/item_upload_files.R:266]: There are some R gql clients out there, e.g. this one: https://github.com/ropensci/ghql.  Maybe they would abstract some of that query construction for you?  I basically do what you're doing here for SQL commands though, it gets the job done.

**Comment by wdwatkins** [comment]
This seems fine to me, though I would not claim to be particularly knowledgable about the package in general or graphql.

File: R/session_renew.R (hunk 10 above)
Comment (dblodgett-usgs) [reply R/session_renew.R:60]: I don't know -- this is the constant from sciencebasepy though.

**Pull request merged**